Add Mini Tutorial to programming guide (#2207)
Xilinx/mlir-aie: programming_guide/mini_tutorial/README.md

# <ins>IRON Mini Tutorial</ins>

**These exercises are set to run on Strix (NPU2) devices. If you wish to run them on a different architecture, you can change the device at the top of the python source files.**

## <ins>Key Components: Workers, ObjectFifos, Runtime</ins>

> IRON 1.0 introduces an unplaced layer to the [IRON API](../../python/iron/). As an example, below are the unplaced versions of the AIE compute code and the Object FIFO data movement primitive:

Reviewer: ```suggestion
IRON provides an unplaced (deferred placement) [API](../../python/iron/) for NPU programming. Below are examples describing AIE compute code and the Object FIFO data movement primitive:
```

# <ins>IRON Mini Tutorial</ins>

> **These exercises are set to run on Strix (NPU2) devices. If you wish to run them on a different architecture, you can change the device at the top of the python source files.**

Reviewer: ```suggestion
**These exercises are set to run on NPU2 (Strix, Strix Halo, Krackan) devices. If you wish to run them on a different architecture, you can change the device at the top of the python source files.**
```

tile_size = data_size // n_workers
tile_ty = np.ndarray[(tile_size,), np.dtype[np.int32]]

# Dataflow with ObjectFifos
of_offsets = [tile_size * worker for worker in range(n_workers)]

of_out = ObjectFifo(data_ty, name="out")
of_outs = of_out.prod().join(
    of_offsets,
    obj_types=[tile_ty] * n_workers,
    names=[f"out{worker}" for worker in range(n_workers)],
)

def core_fn(of_out):
    # ...kernel code...

# Create workers to perform the tasks
workers = []
for worker in range(n_workers):
    workers.append(
        Worker(core_fn, [of_outs[worker].prod()])
    )
```
More on the Object FIFO data movement patterns in [Section 2b](../section-2/section-2b/README.md) of the programming guide.

## <u>Exercises</u>
6. Familiarize yourself with [exercise_2](./exercise_2/aie2.py). Modify the code in [exercise_2](./exercise_2/aie2.py) so that the input data is split between three workers and their outputs are joined before the final result is sent to external memory.

7. Run `make clean`. Modify the code in [exercise_2](./exercise_2/aie2.py) such that the data sizes for each worker are uneven, for example: tile_sizes = [8, 24, 16].

## <ins>Runtime Parameters and Barriers</ins>

> IRON 1.0 also supports the use of Runtime Parameters which are set and propagated to the Workers at runtime.

Reviewer: ```suggestion
IRON supports Runtime Parameters which are set and propagated to the Workers at runtime.
```

```python
n_workers = 4

# Define tensor types
data_size = 256
data_ty = np.ndarray[(data_size,), np.dtype[np.int32]]

# Runtime parameters
rtps = []
for i in range(n_workers):
    rtps.append(
        GlobalBuffer(
            np.ndarray[(16,), np.dtype[np.int32]],
            name=f"rtp{i}",
            use_write_rtp=True,
        )
    )

def core_fn(rtp):
    runtime_parameter = rtp

# Create workers to perform the tasks
workers = []
for worker in range(n_workers):
    workers.append(
        Worker(core_fn, [rtps[worker]])
    )

rt = Runtime()
with rt.sequence(data_ty, data_ty, data_ty) as (_, _, _):

    # Set runtime parameters
    def set_rtps(*args):
        for rtp in args:
            rtp[0] = 50
            rtp[1] = 255
            rtp[2] = 0

    rt.inline_ops(set_rtps, rtps)
```

> To ensure that RTPs are not read prematurely, IRON 1.0 introduces `WorkerRuntimeBarriers` which can be used to synchronize a Worker with the runtime sequence:

Reviewer: ```suggestion
To ensure that RTPs are not read prematurely, `WorkerRuntimeBarriers` can be used to synchronize a Worker with the runtime sequence:
```

9. The design fails because the Worker reads the RTP before the runtime sets it. Modify the code in [exercise_3](./exercise_3/aie2.py) such that the Worker uses a `WorkerRuntimeBarrier` to wait for the RTP to be set.

10. 

## <ins>Advanced Topic: Data Layout Transformations</ins>

AIE array DMAs can perform on-the-fly data transformations. Transformations on each dimension are expressed as a (size, stride) pair. Dimensions are given from highest to lowest:
```python
[(size_2, stride_2), (size_1, stride_1), (size_0, stride_0)]
```

Data layout transformations can be viewed as a way to specify to the hardware which location in the data to access next and as such it is possible to model the access pattern using a series of nested loops. For example, the transformation using the strides and sizes from above can be expressed as:
```c
int *buffer;
for(int i = 0; i < size_2; i++)
    for(int j = 0; j < size_1; j++)
        for(int k = 0; k < size_0; k++)
            // access/store element at/to buffer[  i * stride_2
            //                                   + j * stride_1
            //                                   + k * stride_0]
```

> To better support DMA on-the-fly data transformations **at runtime** IRON 1.0 introduces [taplib](../../python/helpers/taplib/) which provides the building blocks for `Tensor Access Pattern`s (`taps`). The sizes and strides are grouped together and the dimensions should be given from highest to lowest:

Reviewer: ```suggestion
To better support DMA on-the-fly data transformations **at runtime** IRON provides [taplib](../../python/helpers/taplib/) which provides the building blocks for `Tensor Access Pattern`s (`taps`). The sizes and strides are grouped together and the dimensions should be given from highest to lowest:
```
